### PR TITLE
Alt tag method

### PR DESCRIPTION
**Pull request by dakota002**
Alternate method as discussed in #101 

@lpsinger , the only concern I have is that this will create the tag/release on the user's fork. Would the user who triggers the update need to be working in a clone of the nasa-gcn repo specifically for this method to work as intended?

### CODE REVIEW

**Comment by lpsinger** [comment]
> Alternate method as discussed in #101
> 
> @lpsinger , the only concern I have is that this will create the tag/release on the user's fork. Would the user who triggers the update need to be working in a clone of the nasa-gcn repo specifically for this method to work as intended?

Yes. So we should document the release process.

**Comment by dakota002** [comment]
Sounds good, I can update the readme. Will this also require any changes in the repo's settings to allow users to push to main?

File: update-tags.mjs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import { readFile, writeFile } from 'fs/promises'
+import { exec } from 'child_process'
+import { glob as baseGlob } from 'glob'
+import packageJSON from './package.json' assert { type: 'json' }
+
+const tagPath = `/v${packageJSON.version}/`
+
+async function glob(path) {
+  return await baseGlob(path, {
+    ignore: ['test/**', 'node_modules/**'],
```
Comment (lpsinger) [inline update-tags.mjs:10]: The `test` directory does not exist.

```suggestion
    ignore: ['node_modules/**'],
```

Alternatively, could we get the list of patterns to ignore from `.gitignore`?

File: update-tags.mjs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    exampleFilenames.map(async (path) => {
+      const example = JSON.parse(
+        await readFile(path, {
+          encoding: 'utf-8',
+        })
+      )
+      example['$schema'] = example['$schema'].replace(oldValue, newValue)
+      await writeFile(path, JSON.stringify(example, null, 2))
+    })
+  )
```
Comment (lpsinger) [inline update-tags.mjs:38]: Please refactor to eliminate duplicated code.

Also, I am concerned that the search-and-replace method here could unintentionally overwrite substrings that coincidentally match the Git ref --- for example, `foo.bar.main.Bat`. Admittedly, a contrived corner case, but a corner case nonetheless. Please instead test for whether the value _begins with_ a GitHub URL matching the current branch, and if so, replace _the matching portion at the beginning of the string_.

File: update-tags.mjs [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import { readFile, writeFile } from 'fs/promises'
+import { exec } from 'child_process'
+import { glob as baseGlob } from 'glob'
+import packageJSON from './package.json' assert { type: 'json' }
```
Comment (lpsinger) [inline update-tags.mjs:4]: Is the following legal?

```suggestion
import { version } from './package.json' assert { type: 'json' }
```

File: update-tags.mjs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  )
+}
+
+if (process.argv.some((x) => x == '--reset')) {
+  await loadAndUpdateFiles(tagPath, '/main/')
+} else {
+  await loadAndUpdateFiles('/main/', tagPath)
+}
+
+exec('npx prettier -w .')
```
Comment (lpsinger) [inline update-tags.mjs:47]: Instead of running prettier, can we pass arguments to `JSON.stringify` that ensure that it will match Prettier's settings?

File: update-tags.mjs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+          encoding: 'utf-8',
+        })
+      )
+      example['$schema'] = example['$schema'].replace(oldValue, newValue)
+      await writeFile(path, JSON.stringify(example, null, 2))
+    })
+  )
+}
+
+if (process.argv.some((x) => x == '--reset')) {
```
Comment (lpsinger) [inline update-tags.mjs:41]:
```suggestion
if (process.argv.includes('--reset')) {
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+      - '*'
+
+jobs:
+  tag-new-versions:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+    steps:
+      - uses: actions/checkout@v3
+      - uses: ncipollo/release-action@v1
```
[the remark this answers is not in the document]
Comment (dakota002) [reply .github/workflows/release.yml:14]: I'll look into some alternatives

File: README.md [hunk 7]
```diff
@@ -8,3 +8,12 @@ Please add your schema to this repository under <code>gcn/notices/<i>mission</i>
...
 The GCN team is happy to iterate with the producers on their schema contents and format.
 Your pipeline will generate JSON files following these schema and send alerts to GCN as described in [New Notice Producers](https://gcn.nasa.gov/docs/producers).
+
+## Version Updates
+
+New versions are simplified with the [`npm version`](https://docs.npmjs.com/cli/v8/commands/npm-version) command. The following steps describe the process for incrementing versions.
+
+1. Clone the parent repo (nasa-gcn/gcn-schema). This should not be run in your own forked version of this repo.
+
+2. Run `npm version [ major | minor | patch ]`
```
[the remark this answers is not in the document]
Comment (dakota002) [reply README.md:18]: The push commands are tied into the postversion script which runs automatically with the ncipollo/release-action@v1 action, but if we change this it will need to be updated

File: update-tags.mjs (hunk 3 above)
Comment (dakota002) [reply update-tags.mjs:4]: This threw errors when I tried to implement it, something along the lines of package.json not having a valid export named version

File: update-tags.mjs (hunk 4 above)
Comment (dakota002) [reply update-tags.mjs:47]: Mostly, though it seems like the one specific that is getting in the way is array formatting:

json.stringify will output an array like this:

```
   "enum": [
        "circle",
        "ellipse",
        "rectangle"
      ],
```

where prettier parses it as short enough for one line:

```
"enum": ["circle", "ellipse", "rectangle"],
```

File: update-tags.mjs (hunk 3 above)
Comment (lpsinger) [reply update-tags.mjs:4]: What errors?

File: update-tags.mjs (hunk 4 above)
Comment (lpsinger) [reply update-tags.mjs:42]: OK. Instead of calling it from the CLI, would you please use the API? https://prettier.io/docs/en/api.html

File: update-tags.mjs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+async function fileUpdates(fileSet, key, oldValue, newValue) {
+  await Promise.all(
+    fileSet.map(async (fileItem) => {
+      let file = JSON.parse(
+        await readFile(fileItem, {
+          encoding: 'utf-8',
+        })
+      )
+      file[key] = file[key].replace(oldValue, newValue)
+      await writeFile(fileItem, JSON.stringify(file, null, 2).concat('\n'))
```
Comment (lpsinger) [inline update-tags.mjs:23]: No need to adjust the formatting if you do have to run it through prettier.

```suggestion
      await writeFile(fileItem, JSON.stringify(file))
```

File: update-tags.mjs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+  const exampleFilenames = await glob('**/*.example.json')
+  await fileUpdates(exampleFilenames, '$schema', oldValue, newValue)
+}
+
+if (process.argv.includes('--reset')) {
+  await loadAndUpdateFiles(tagPath, '/main/')
+} else {
+  await loadAndUpdateFiles('/main/', tagPath)
+}
```
Comment (lpsinger) [inline update-tags.mjs:40]: These regexes admit false positive matches. The global search and replace still has a corner case that it doesn't properly handle paths that _happen_ to contain strings that match version tags or branch names but aren't.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: update-tags.mjs (hunk 3 above)
Comment (dakota002) [reply update-tags.mjs:4]:
```
import { version } from './package.json' assert { type: 'json' }
         ^^^^^^^
SyntaxError: The requested module './package.json' does not provide an export named 'version'
    at ModuleJob._instantiate (node:internal/modules/esm/module_job:123:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:189:5)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:530:24)
    at async loadESM (node:internal/process/esm_loader:91:5)
    at async handleMainPromise (node:internal/modules/run_main:65:12)
```

Comment (lpsinger) [reply update-tags.mjs:4]: Oh, but of course! There is no `version` key in your `package.json` file. So add one!

File: update-tags.mjs [hunk 10]
```diff
@@ -0,0 +1,47 @@
+import { readFile, writeFile } from 'fs/promises'
+import { exec } from 'child_process'
```
[the remark this answers is not in the document]
Comment (dakota002) [reply update-tags.mjs:2]: Not directly, however it does support the following as a means to convert it to a promise:

```
const util = require('node:util');
const exec = util.promisify(require('node:child_process').exec);

async function lsExample() {
  const { stdout, stderr } = await exec('ls');
  console.log('stdout:', stdout);
  console.error('stderr:', stderr);
}
lsExample();
```

I could implement it in a similar way, but if I make use of the prettier api call this will go away

File: .github/workflows/release.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (dakota002) [reply .github/workflows/release.yml:14]: Still reviewing alternate options, but [gh-release](https://github.com/marketplace/actions/gh-release) looks like a potential replacement.

File: .github/workflows/release.yml [hunk 11]
```diff
@@ -0,0 +1,14 @@
+name: Create New Release
+on:
+  push:
+    tags:
+      - '*'
```
[the remark this answers is not in the document]
Comment (dakota002) [reply .github/workflows/release.yml:5]: I think this is the most generic for allowing any tag, the alternative being '**' https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#using-event-activity-types

File: update-tags.mjs (hunk 3 above)
Comment (dakota002) [reply update-tags.mjs:3]: Unfortunately that error happens with a `version` key as well:
image

File: update-tags.mjs (hunk 9 above)
Comment (dakota002) [reply update-tags.mjs:43]: True, I have updated the matching (lines 22-25)

File: package.json [hunk 12]
```diff
@@ -4,7 +4,9 @@
     "prepare": "husky install",
     "validate:prettier": "prettier --check .",
     "validate:jsonschema": "node validate.mjs",
-    "validate": "run-s validate:*"
+    "validate": "run-s validate:*",
+    "version": "node update-tags.mjs && git add -u",
+    "postversion": "node update-tags.mjs --reset && git add -u && git commit -m \"Back to development\" && git push && git push --tags"
```
[the remark this answers is not in the document]
Comment (dakota002) [reply package.json:9]: Do you think it makes sense then to remove the `postversion` command all together?  Or maybe it makes sense to add a command that could be run, like `npm review`, that can display the new version tag and any other possibly relevant information?

[the remark this answers is not in the document]
Comment (lpsinger) [reply package.json:9]: > Do you think it makes sense then to remove the `postversion` command all together?

No, you still need the "Back to development" commit.

> Or maybe it makes sense to add a command that could be run, like `npm review`, that can display the new version tag and any other possibly relevant information?

No. I think that the instructions should just tell the developer to review at the changes with `git log -p` before pushing them.

File: update-tags.mjs (hunk 3 above)
Comment (lpsinger) [reply update-tags.mjs:3]: How interesting! Well, thanks anyway for trying.

File: .github/workflows/release.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (dakota002) [reply .github/workflows/release.yml:14]: What are your thoughts on using [action-gh-release](https://github.com/softprops/action-gh-release)? 
Almost 3k stars and fewer files than the other one I was checking out.

[the remark this answers is not in the document]
Comment (lpsinger) [reply .github/workflows/release.yml:14]: I like it. Did you audit the code too?

**Comment by lpsinger** [comment]
Remember to re-request a review when you are ready for me to look at this again.

File: .github/workflows/release.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (dakota002) [reply .github/workflows/release.yml:14]: I did read over the code, it looks like its as straight forward as it gets for this

File: README.md [hunk 13]
```diff
@@ -8,3 +8,18 @@ Please add your schema to this repository under <code>gcn/notices/<i>mission</i>
...
+   `git clone git@github.com:nasa-gcn/gcn-schema.git `
+
+2. Run `npm version [ major | minor | patch ]`
+   This command will handle the intermediate steps of updating and committing the path changes in each file as defined in the `version` and `postversion` npm scripts
+
+3. Review the changes with `git log -p` to make sure that each file is appropriately updated.
+
+4. Push the commit and tag:
+
+   `git push && git push origin <tag name>`
```
Comment (lpsinger) [inline README.md:25]:
```suggestion
1.  Clone the parent repo (nasa-gcn/gcn-schema).

        git clone git@github.com:nasa-gcn/gcn-schema.git

2.  Run `npm version [ major | minor | patch ]`

    This command will handle the intermediate steps of updating and committing the path changes in each file as defined in the `version` and `postversion` npm scripts

3.  Review the changes with `git log -p` to make sure that each file is appropriately updated.

4.  Push the commit and tag:

    git push && git push origin <tag name>
```

File: update-tags.mjs [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+for (const file of files) {
+  const text = await readFile(file, 'utf-8')
+  await writeFile(
+    file,
+    prettier.format(text, {
+      semi: false,
+      singleQuote: true,
+      parser: 'json',
+    })
+  )
```
Comment (lpsinger) [inline update-tags.mjs:55]: Run prettier inside `loadAndUpdateFiles` so that we don't read and write the same files multiple times.

File: update-tags.mjs [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  await fileUpdates(schemaFilenames, '$id', oldValue, newValue)
+
+  const exampleFilenames = await glob('**/*.example.json')
+  await fileUpdates(exampleFilenames, '$schema', oldValue, newValue)
+}
+
+if (process.argv.includes('--reset')) {
+  await loadAndUpdateFiles(tagPath, '/main/')
+} else {
+  await loadAndUpdateFiles('/main/', tagPath)
```
Comment (lpsinger) [inline update-tags.mjs:42]:
```suggestion
  await loadAndUpdateFiles(tagPath, 'main')
} else {
  await loadAndUpdateFiles('main', tagPath)
```

File: update-tags.mjs [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+  await Promise.all(
+    fileSet.map(async (fileItem) => {
+      let file = JSON.parse(
+        await readFile(fileItem, {
+          encoding: 'utf-8',
+        })
+      )
+      file[key] = file[key].replace(
+        `https://gcn.nasa.gov/schema/${oldValue}`,
+        `https://gcn.nasa.gov/schema/${newValue}`
```
Comment (lpsinger) [inline update-tags.mjs:24]:
```suggestion
        `https://gcn.nasa.gov/schema/${oldValue}/`,
        `https://gcn.nasa.gov/schema/${newValue}/`
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: update-tags.mjs [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+      file[key] = file[key].replace(
+        `https://gcn.nasa.gov/schema/${oldValue}/`,
+        `https://gcn.nasa.gov/schema/${newValue}/`
+      )
+
+      await writeFile(
+        fileItem,
+        prettier.format(JSON.stringify(file), {
+          semi: false,
+          singleQuote: true,
```
Comment (lpsinger) [inline update-tags.mjs:31]: I doubt that either the `semi` or `singleQuote` options have any effect on JSON documents.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: update-tags.mjs (hunk 17 above)
Comment (dakota002) [reply update-tags.mjs]: Yea, just removed them

**Review by lpsinger — APPROVED** [review]
